// @temando/serverless-temando-registry/dist/CfVariables.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

require('babel-polyfill');

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _fsPromise = require('fs-promise');

var _fsPromise2 = _interopRequireDefault(_fsPromise);

var _lutils = require('lutils');

var _jsYaml = require('js-yaml');

var _jsYaml2 = _interopRequireDefault(_jsYaml);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Enumerates `keys`, reducing a pointer in `data`
 * Consumes undefined or bad paths silently
 *
 * @param {Array} keys
 * @param {Object} data
 * @return {any}
 * @return {undefined} When invalid pointer
 */
function getPointer(keys, data) {
  return keys.reduce(function (pointer, key) {
    if (_lutils.typeOf.Object(pointer) || _lutils.typeOf.Array(pointer)) {
      return pointer[key];
    }

    return undefined;
  }, data);
}

var CfVariables = function () {
  function CfVariables(_ref) {
    var yaml = _ref.yaml,
        aws = _ref.aws,
        _ref$storagePath = _ref.storagePath,
        storagePath = _ref$storagePath === undefined ? './serverless.env.yml' : _ref$storagePath,
        servicePath = _ref.servicePath,
        region = _ref.region,
        stage = _ref.stage,
        _ref$log = _ref.log,
        log = _ref$log === undefined ? function () {} : _ref$log;
    (0, _classCallCheck3.default)(this, CfVariables);

    this.yaml = yaml;
    this.aws = aws;

    this.region = region;
    this.stage = stage;

    this.storagePath = _path2.default.resolve(servicePath, storagePath);
    this.slsYmlPath = _path2.default.resolve(servicePath, 'serverless.yml');

    this.varPath = 'cf';
    this.envKey = 'cf-environment';

    this.log = log;
  }

  (0, _createClass3.default)(CfVariables, [{
    key: 'initialize',
    value: function () {
      var _ref2 = (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee() {
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this.initialized) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt('return');

              case 2:
                _context.next = 4;
                return this.parse();

              case 4:
                _context.next = 6;
                return this.loadStore();

              case 6:

                this.initialized = true;

              case 7:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function initialize() {
        return _ref2.apply(this, arguments);
      }

      return initialize;
    }()

    /**
     * Parses cf-environment mappings from `serverless.yml`
     *
     * @returns this
     */

  }, {
    key: 'parse',
    value: function () {
      var _ref3 = (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee2() {
        var _this = this;

        var mappings = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];

        var slsYml, providerEnv, _mappings;

        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                this.mappings = mappings;

                _context2.next = 3;
                return this.yaml.parse(this.slsYmlPath);

              case 3:
                slsYml = _context2.sent;
                providerEnv = slsYml.provider && slsYml.provider[this.envKey];


                if (providerEnv) {
                  (_mappings = this.mappings).push.apply(_mappings, (0, _toConsumableArray3.default)(this.parseEnvMap(providerEnv, { definedAt: ['provider'] })));
                }

                if (slsYml.functions) {
                  (0, _keys2.default)(slsYml.functions).forEach(function (fnKey) {
                    var fnConfig = slsYml.functions[fnKey];
                    var fnEnv = fnConfig && fnConfig[_this.envKey];

                    if (fnEnv) {
                      var _mappings2;

                      (_mappings2 = _this.mappings).push.apply(_mappings2, (0, _toConsumableArray3.default)(_this.parseEnvMap(fnEnv, { definedAt: ['functions', fnKey] })));
                    }
                  });
                }

                return _context2.abrupt('return', this);

              case 8:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function parse() {
        return _ref3.apply(this, arguments);
      }

      return parse;
    }()
  }, {
    key: 'parseEnvMap',
    value: function parseEnvMap(envMap, mappingTemplate) {
      return envMap.map(function (definition) {
        var mapping = (0, _extends3.default)({}, mappingTemplate);

        if (_lutils.typeOf.String(definition)) {
          /**
           * cf-environment:
           *   - myKeyName
           */

          mapping.cfKey = definition;
          mapping.key = definition;
        } else if (_lutils.typeOf.Object(definition)) {
          /**
           * cf-environment:
           *   - FOO: bar
           */

          var firstKey = (0, _keys2.default)(definition)[0];
          mapping.cfKey = definition[firstKey];
          mapping.key = firstKey;
        } else {
          throw new Error('Invalid cf definition type - must be a string or dict');
        }

        return mapping;
      });
    }

    /**
     * Loads in variables from a storage file
     */

  }, {
    key: 'loadStore',
    value: function () {
      var _ref4 = (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee3() {
        var store, definedAt, selected;
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                store = void 0;

                if (_fsPromise2.default.existsSync(this.storagePath)) {
                  _context3.next = 5;
                  break;
                }

                store = {};
                _context3.next = 8;
                break;

              case 5:
                _context3.next = 7;
                return this.yaml.parse(this.storagePath);

              case 7:
                store = _context3.sent;

              case 8:
                definedAt = [this.varPath, this.stage, this.region];

                // Used to ensure store maintains reference to `selected` for updates

                selected = definedAt.reduce(function (pointer, key) {
                  if (!pointer[key]) pointer[key] = {};
                  return pointer[key];
                }, store);


                this.store = {
                  definedAt,
                  data: store,

                  // Selected will be at `cf[stage][region]` {}
                  selected
                };

                return _context3.abrupt('return', this);

              case 12:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function loadStore() {
        return _ref4.apply(this, arguments);
      }

      return loadStore;
    }()
  }, {
    key: 'saveStore',
    value: function () {
      var _ref5 = (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee4() {
        var stringified;
        return _regenerator2.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                stringified = _jsYaml2.default.dump(this.store.data);
                return _context4.abrupt('return', _fsPromise2.default.writeFile(this.storagePath, stringified));

              case 2:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function saveStore() {
        return _ref5.apply(this, arguments);
      }

      return saveStore;
    }()
  }, {
    key: 'updateStore',
    value: function updateStore(data) {
      return (0, _lutils.merge)(this.store.selected, data);
    }
  }, {
    key: 'getStoreVar',
    value: function getStoreVar(cfKey) {
      return this.store.selected[cfKey];
    }
  }, {
    key: 'populateEnvironments',
    value: function populateEnvironments(config) {
      var _this2 = this;

      this.mappings.forEach(function (_ref6) {
        var definedAt = _ref6.definedAt,
            key = _ref6.key,
            cfKey = _ref6.cfKey;

        var parent = getPointer(definedAt, config);
        parent.environment = parent.environment || {};
        parent.environment[key] = _this2.getStoreVar(cfKey);
      });
    }
  }, {
    key: 'updateLambdaEnvironments',
    value: function () {
      var _ref7 = (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee6(config) {
        var _this3 = this;

        var cfVars = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.store.selected;
        return _regenerator2.default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return _bluebird2.default.map((0, _keys2.default)(config.functions), function () {
                  var _ref8 = (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee5(fnName) {
                    var stored, mappedEnv, baseEnvironment, environment;
                    return _regenerator2.default.wrap(function _callee5$(_context5) {
                      while (1) {
                        switch (_context5.prev = _context5.next) {
                          case 0:
                            stored = {};
                            mappedEnv = _this3.mappings.reduce(function (obj, _ref9) {
                              var definedAt = _ref9.definedAt,
                                  key = _ref9.key,
                                  cfKey = _ref9.cfKey;

                              if (definedAt[0] === 'functions' && definedAt[1] === fnName || definedAt[0] === 'provider') {
                                stored[key] = _this3.getStoreVar(cfKey);
                                obj[key] = cfVars[cfKey];
                              }

                              return obj;
                            }, {});
                            baseEnvironment = (0, _lutils.merge)((0, _lutils.clone)(config.provider.environment || {}), (0, _lutils.clone)(config.functions[fnName].environment || {}));
                            environment = (0, _lutils.merge)((0, _lutils.clone)(baseEnvironment || {}), mappedEnv);

                            /**
                             * Determines if there have been any changes
                             */

                            if (!((0, _stringify2.default)(environment) === (0, _stringify2.default)(baseEnvironment))) {
                              _context5.next = 6;
                              break;
                            }

                            return _context5.abrupt('return');

                          case 6:

                            _this3.log(`[ENVIRONMENT] Updating deployed function '${fnName}' environment`);

                            _context5.next = 9;
                            return _this3.aws.request('Lambda', 'updateFunctionConfiguration', {
                              FunctionName: config.functions[fnName].name,
                              Environment: {
                                Variables: environment
                              }
                            });

                          case 9:
                          case 'end':
                            return _context5.stop();
                        }
                      }
                    }, _callee5, _this3);
                  }));

                  return function (_x4) {
                    return _ref8.apply(this, arguments);
                  };
                }());

              case 2:
              case 'end':
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function updateLambdaEnvironments(_x2) {
        return _ref7.apply(this, arguments);
      }

      return updateLambdaEnvironments;
    }()
  }]);
  return CfVariables;
}();

exports.default = CfVariables;
module.exports = exports['default'];
