// @temando/service-library/schema/filter-fields-config-schema+v1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const operatorMap_1 = require("./../middlewares/SearchFilter/operatorMap");
// tslint:disable-next-line:no-default-export
exports.default = {
    $schema: 'http://json-schema.org/draft-04/schema#',
    type: 'object',
    id: 'https://schemas.temando.com/filter-fields-config-schema+v1.json',
    description: 'Set of filterable fields and their internal mapping and supported operators.',
    additionalProperties: {
        $ref: '#/definitions/filterField',
    },
    definitions: {
        filterField: {
            type: 'object',
            description: 'Field internal mapping and supported operators.',
            required: [
                'mappedPath',
                'supportedOperators',
            ],
            properties: {
                mappedPath: {
                    type: 'string',
                    description: 'The document path to the stored field. Prefixed by `/Attributes/`.',
                    pattern: /^\/Attributes\/[\x00-\x7F]*[a-zA-Z0-9/]+$/.source,
                },
                supportedOperators: {
                    type: 'array',
                    description: 'A set of valid operators that can be used for this filterable field.',
                    items: {
                        type: 'string',
                        enum: Object.keys(operatorMap_1.default),
                    },
                },
            },
        },
    },
};
