// @temando/generic-schema/schemas/address.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
exports.default = {
    $schema: 'http://json-schema.org/draft-04/schema#',
    id: configuration_1.default.prefix + "/address#",
    description: 'Address.',
    type: 'object',
    additionalProperties: false,
    required: [
        'countryCode',
    ],
    properties: {
        lines: {
            description: 'One or more address lines. For example:' +
                '\n- Street name and number' +
                '\n- Floor or suite number.',
            type: 'array',
            items: {
                type: 'string',
            },
        },
        countryCode: {
            $ref: configuration_1.default.prefix + "/country#",
        },
        postalCode: {
            type: 'string',
            description: 'Address postal code (for example, ZIP code or post code).',
        },
        locality: {
            type: 'string',
            description: 'Address locality. For example city or township.\n\nSome countries do not use this field and ' +
                'address lines are sufficient to locate the address within the administrative area.',
        },
        type: {
            type: 'string',
            description: 'The type of property at the address.',
            enum: [
                'business',
                'residential',
                'post-office-box',
                'collection-point',
            ],
        },
        dependentLocality: {
            type: 'string',
            description: 'Address dependent locality (neighbourhood).\n\nWhen representing a ' +
                'double-dependent locality in Great Britain, this includes both the double-dependent locality and ' +
                'the dependent locality,',
        },
        administrativeArea: {
            type: 'string',
            description: 'Address administrative area. For example, state or province.',
        },
        longitude: {
            type: 'number',
            description: 'The longitude value of a geo-coded address.',
            minimum: -180,
            maximum: 180,
        },
        latitude: {
            type: 'number',
            description: 'The latitude value of a geo-coded address.',
            minimum: -90,
            maximum: 90,
        },
    },
};
