// @temando/generic-schema/schemas/capabilities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
exports.default = {
    $schema: 'http://json-schema.org/draft-04/schema#',
    id: configuration_1.default.prefix + "/capabilities#",
    description: 'Carrier service capabilities and options.',
    type: 'object',
    additionalProperties: false,
    properties: {
        authorityToLeave: {
            $ref: '#/definitions/authorityToLeave',
        },
        signature: {
            $ref: '#/definitions/signatureOnDelivery',
        },
    },
    definitions: {
        authorityToLeave: {
            description: 'Authority to leave.',
            type: 'object',
            properties: {
                required: {
                    type: 'boolean',
                },
            },
        },
        signatureOnDelivery: {
            description: 'Signature on delivery.',
            type: 'object',
            required: [
                'required',
            ],
            properties: {
                required: {
                    type: 'boolean',
                },
                onlyAdultCanSign: {
                    type: 'boolean',
                },
                onlyAddresseeCanSign: {
                    type: 'boolean',
                },
            },
        },
    },
};
