// @temando/generic-schema/schemas/dimensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
exports.default = {
    $schema: 'http://json-schema.org/draft-04/schema#',
    id: configuration_1.default.prefix + "/dimensions#",
    description: 'Dimensions.',
    type: 'object',
    additionalProperties: false,
    required: [
        'length',
        'width',
        'height',
        'unit',
    ],
    properties: {
        length: {
            type: 'number',
            minimum: 0,
        },
        width: {
            type: 'number',
            minimum: 0,
        },
        height: {
            type: 'number',
            minimum: 0,
        },
        unit: {
            description: 'Applicable unit of measure.',
            type: 'string',
            enum: [
                'mm',
                'millimeter',
                'millimetre',
                'cm',
                'centimeter',
                'centimetre',
                'm',
                'meter',
                'metre',
                'ft',
                'foot',
                'feet',
                'in',
                'inch',
                'yd',
                'yard',
            ],
        },
    },
};
