// @temando/generic-schema/schemas/document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
exports.default = {
    $schema: 'http://json-schema.org/draft-04/schema#',
    id: configuration_1.default.prefix + "/document#",
    description: 'Document.',
    additionalProperties: false,
    type: 'object',
    required: [
        'size',
        'type',
        'mimeType',
    ],
    properties: {
        size: {
            $ref: configuration_1.default.prefix + "/document-size#",
        },
        type: {
            $ref: configuration_1.default.prefix + "/document-type#",
        },
        mimeType: {
            description: 'Document-specific MIME type.',
            $ref: configuration_1.default.prefix + "/mime-type#",
            // Document-specific validation override
            enum: [
                'application/epl',
                'application/epl2',
                'application/pdf',
                'application/zpl',
                'image/bmp',
                'image/gif',
                'image/jpg',
                'image/png',
                'text/html',
                'text/plain',
            ],
        },
        url: {
            description: 'URL from which the document can be retrieved.',
            type: 'string',
            format: 'uri',
        },
        data: {
            description: 'Document content as a base64-encoded string.',
            type: 'string',
        },
        encoding: {
            $ref: configuration_1.default.prefix + "/encoding#",
        },
    },
};
