// @temando/generic-schema/schemas/entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
exports.default = {
    id: configuration_1.default.prefix + "/entity#",
    $schema: 'http://json-schema.org/draft-04/schema#',
    description: 'Entity.',
    type: 'object',
    additionalProperties: false,
    properties: {
        organisationName: {
            type: 'string',
            description: 'Entity\'s organisation name.',
        },
        personTitle: {
            type: 'string',
            description: 'Title of entity\'s contact. For example, Mr, Mrs, or Ms.',
        },
        personFirstName: {
            type: 'string',
            description: 'First name of entity\'s contact.',
        },
        personLastName: {
            type: 'string',
            description: 'Last name of entity\'s contact.',
        },
        email: {
            type: 'string',
            description: 'Entity\'s contact email address.',
            format: 'email',
        },
        phoneNumber: {
            type: 'string',
            description: 'Entity\'s contact phone number.',
        },
        secondaryPhoneNumber: {
            type: 'string',
            description: 'Entity\'s alternate contact phone number.',
        },
        mobilePhoneNumber: {
            type: 'string',
            description: 'Entity\'s contact mobile phone number.',
        },
        faxNumber: {
            type: 'string',
            description: 'Entity\'s contact fax number.',
        },
        nationalIdentificationNumber: {
            type: 'string',
            description: 'National identification number of this entity. For example, the INSEE code (France).',
        },
        taxIdentificationNumber: {
            type: 'string',
            description: 'Tax identification number of this entity. For example, TFN (Australia) or TIN (France).',
        },
    },
};
