// @temando/generic-schema/schemas/experience.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
exports.default = {
    description: 'Experience.',
    $schema: 'http://json-schema.org/draft-04/schema#',
    id: configuration_1.default.prefix + "/experience#",
    additionalProperties: false,
    type: 'object',
    required: [
        'code',
    ],
    properties: {
        code: {
            type: 'string',
            description: 'Experience unique ID.',
        },
        costs: {
            type: 'array',
            items: {
                description: 'Experience cost.',
                $ref: configuration_1.default.prefix + "/money#",
            },
        },
        descriptions: {
            type: 'array',
            items: {
                description: 'Experience description.',
                $ref: "#/definitions/description",
            },
        },
    },
    definitions: {
        description: {
            type: 'object',
            required: [
                'language',
                'text',
            ],
            properties: {
                locale: {
                    type: 'string',
                    description: 'BCP-47 code identifying language of description text.',
                },
                text: {
                    type: 'string',
                    description: 'Description text.',
                },
            },
        },
    },
};
