// @temando/generic-schema/schemas/exportDeclaration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
exports.default = {
    $schema: 'http://json-schema.org/draft-04/schema#',
    id: configuration_1.default.prefix + "/exportDeclaration#",
    description: 'Shipment export declaration.',
    type: 'object',
    additionalProperties: false,
    properties: {
        signatory: {
            description: 'Shipment\'s export declaration signatory.',
            $ref: configuration_1.default.prefix + "/entity#",
        },
        exportReason: {
            description: 'Description of reason for export.',
            type: 'string',
        },
        exportCategory: {
            description: 'Category of export. For example, commercial, documents, gift, or other.',
            type: 'string',
        },
        exportCodes: {
            type: 'object',
            description: 'Export codes associated with the shipment.',
            properties: {
                exportDeclarationNumber: {
                    type: 'string',
                    description: 'Export Declaration Number (EDN) for exports originating from Australia.',
                    minLength: 0,
                },
                exemptionExclusionLegend: {
                    type: 'string',
                    description: 'Exemption Exclusion Legend (EEL) for exports originating from the United States.',
                    minLength: 0,
                },
                electronicExportInformation: {
                    type: 'string',
                    description: 'Electronic Export Information (EEI), formerly known as Shipper\'s ' +
                        'Export Declaration (SED), for exports originating from the United States.',
                    minLength: 0,
                },
                internalTransactionNumber: {
                    type: 'string',
                    description: 'Internal Transaction Number (ITN) for exports originating from the United States.' +
                        'A generated number confirming that the EEI was accepted and is on file.',
                    minLength: 0,
                },
            },
        },
        declaredValue: {
            $ref: configuration_1.default.prefix + "/money#",
        },
        descriptionOfGoods: {
            description: 'Description of goods being shipped.',
            type: 'string',
        },
        incoterm: {
            description: 'Applicable trade term. Valid terms are from Incoterm 2010 rules.',
            enum: [
                'EXW',
                'FCA',
                'FAS',
                'FOB',
                'CPT',
                'CFR',
                'CIF',
                'CIP',
                'DAT',
                'DAP',
                'DDP',
            ],
        },
    },
};
