// @temando/generic-schema/schemas/filter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
exports.default = {
    $schema: 'http://json-schema.org/draft-04/schema#',
    id: configuration_1.default.prefix + "/filter",
    description: 'Filtering equation for listing resources.',
    type: 'array',
    items: {
        type: 'object',
        required: [
            'path',
            'operator',
            'value',
        ],
        properties: {
            path: {
                description: 'Valid JSON pointer to the document type to filter on relative to the `/data/attributes/` path from a ' +
                    'standard response. For example, given a response with `/data/attributes/origin/address/countryCode` the path ' +
                    'is `origin/address/countryCode`.',
                type: 'string',
                pattern: /^\/[\x00-\x7F]*[a-zA-Z0-9]$/.source,
            },
            operator: {
                description: 'Equation operator.',
                type: 'string',
                enum: [
                    'equal',
                    'notEqual',
                    'greaterThan',
                    'lessThan',
                    'greaterThanOrEqual',
                    'lessThanOrEqual',
                    'in',
                    'notIn',
                    'beginsWith',
                    'exists',
                    'contains',
                    'doesNotContain',
                ],
            },
            value: {
                description: 'Filter value. Forms the right operand of a filter equation.',
                type: [
                    'string',
                    'array',
                    'boolean',
                    'number',
                    'null',
                ],
                minLength: 1,
            },
        },
    },
};
