// @temando/generic-schema/schemas/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var address_1 = require("./address");
var capabilities_1 = require("./capabilities");
var country_1 = require("./country");
var dateRange_1 = require("./dateRange");
var dimensions_1 = require("./dimensions");
var document_1 = require("./document");
var documentSize_1 = require("./documentSize");
var documentType_1 = require("./documentType");
var encoding_1 = require("./encoding");
var entity_1 = require("./entity");
var experience_1 = require("./experience");
var exportCategory_1 = require("./exportCategory");
var exportDeclaration_1 = require("./exportDeclaration");
var filter_1 = require("./filter");
var fulfillmentType_1 = require("./fulfillmentType");
var item_1 = require("./item");
var jsonApiErrors_1 = require("./jsonApiErrors");
var keyValueObject_1 = require("./keyValueObject");
var location_1 = require("./location");
var mimeType_1 = require("./mimeType");
var money_1 = require("./money");
var openingHours_1 = require("./openingHours");
var package_1 = require("./package");
var packagingType_1 = require("./packagingType");
var product_1 = require("./product");
var qualification_1 = require("./qualification");
var quote_1 = require("./quote");
var quoteSet_1 = require("./quoteSet");
var shipment_1 = require("./shipment");
var time_1 = require("./time");
var timeDuration_1 = require("./timeDuration");
var uuid_1 = require("./uuid");
var volume_1 = require("./volume");
var weight_1 = require("./weight");
var zone_1 = require("./zone");
exports.default = {
    address: address_1.default,
    capabilities: capabilities_1.default,
    country: country_1.default,
    dateRange: dateRange_1.default,
    dimensions: dimensions_1.default,
    document: document_1.default,
    documentSize: documentSize_1.default,
    documentType: documentType_1.default,
    encoding: encoding_1.default,
    entity: entity_1.default,
    experience: experience_1.default,
    exportCategory: exportCategory_1.default,
    exportDeclaration: exportDeclaration_1.default,
    filter: filter_1.default,
    fulfillmentType: fulfillmentType_1.default,
    item: item_1.default,
    jsonApiErrors: jsonApiErrors_1.default,
    keyValueObject: keyValueObject_1.default,
    location: location_1.default,
    mimeType: mimeType_1.default,
    money: money_1.default,
    openingHours: openingHours_1.default,
    package: package_1.default,
    packagingType: packagingType_1.default,
    product: product_1.default,
    qualification: qualification_1.default,
    quoteIneligible: quote_1.default.ineligible,
    quoteEligible: quote_1.default.eligible,
    quoteError: quote_1.default.error,
    quoteSet: quoteSet_1.default,
    shipment: shipment_1.default,
    time: time_1.default,
    timeDuration: timeDuration_1.default,
    uuid: uuid_1.default,
    volume: volume_1.default,
    weight: weight_1.default,
    zone: zone_1.default,
};
