// @temando/generic-schema/schemas/jsonApiErrors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
exports.default = {
    $schema: 'http://json-schema.org/draft-04/schema',
    id: configuration_1.default.prefix + "/json-api-errors#",
    description: 'JSON API errors.',
    type: 'object',
    additionalProperties: false,
    properties: {
        errors: {
            type: 'array',
            items: {
                description: 'A JSON API error object.',
                type: 'object',
                properties: {
                    id: {
                        description: 'Error ID.',
                        type: 'string',
                    },
                    status: {
                        description: 'Error status.',
                        type: 'string',
                    },
                    title: {
                        description: 'Error title.',
                        type: 'string',
                    },
                    detail: {
                        description: 'Error details.',
                        type: 'string',
                    },
                    meta: {
                        description: 'Information about the error.',
                        type: 'object',
                    },
                    code: {
                        description: 'Error code.',
                        type: 'string',
                    },
                    links: {
                        description: 'Error links.',
                        type: 'object',
                    },
                    source: {
                        description: 'Error source.',
                        type: 'object',
                    },
                },
            },
        },
    },
};
