// @temando/generic-schema/schemas/location.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
exports.default = {
    $schema: 'http://json-schema.org/draft-04/schema#',
    id: configuration_1.default.prefix + "/location#",
    description: 'Location.',
    type: 'object',
    additionalProperties: false,
    properties: {
        type: {
            description: 'Location type.',
            type: 'string',
            enum: [
                'Store',
                'Warehouse',
                'Drop Shipper',
                'Headquarters',
            ],
        },
        name: {
            description: 'Location name. For example, merchant location name, or collection point reference.\n\nServes as unique ' +
                'identifier for the location.',
            type: 'string',
        },
        address: {
            description: 'Location address.',
            $ref: configuration_1.default.prefix + "/address#",
        },
        contact: {
            description: 'Location contact.',
            $ref: configuration_1.default.prefix + "/entity#",
        },
        openingHours: {
            description: 'Location opening hours.',
            $ref: configuration_1.default.prefix + "/opening-hours#",
        },
    },
};
