// @temando/generic-schema/schemas/openingHours.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
exports.default = {
    $schema: 'http://json-schema.org/draft-04/schema#',
    id: configuration_1.default.prefix + "/opening-hours#",
    description: 'Location opening hours.',
    additionalProperties: false,
    type: 'object',
    required: [
        'default',
    ],
    properties: {
        default: {
            type: 'array',
            items: {
                description: 'Standard opening hours for location.',
                $ref: "#/definitions/dailyOpeningHours",
            },
        },
        exceptions: {
            description: 'Special opening hours for location. For example, holiday periods.',
            additionalProperties: false,
            type: 'object',
            properties: {
                closures: {
                    description: 'Date-time ranges during which the location is closed.',
                    type: 'array',
                    items: {
                        $ref: configuration_1.default.prefix + "/date-range#",
                    },
                },
                openings: {
                    description: 'Date-time ranges during which the location is open outside standard opening hours.',
                    type: 'array',
                    items: {
                        $ref: configuration_1.default.prefix + "/date-range#",
                    },
                },
            },
        },
    },
    definitions: {
        dailyOpeningHours: {
            type: 'object',
            additionalProperties: false,
            required: [
                'closes',
                'dayOfWeek',
                'opens',
            ],
            properties: {
                closes: {
                    $ref: configuration_1.default.prefix + "/time#",
                },
                dayOfWeek: {
                    description: 'Day of the week.',
                    type: 'string',
                    enum: [
                        'monday',
                        'tuesday',
                        'wednesday',
                        'thursday',
                        'friday',
                        'saturday',
                        'sunday',
                    ],
                },
                opens: {
                    $ref: configuration_1.default.prefix + "/time#",
                },
            },
        },
    },
};
