// @temando/generic-schema/schemas/package.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
var dimensions_1 = require("./dimensions");
var document_1 = require("./document");
var item_1 = require("./item");
var packagingType_1 = require("./packagingType");
var volume_1 = require("./volume");
var weight_1 = require("./weight");
exports.default = {
    $schema: 'http://json-schema.org/draft-04/schema#',
    id: configuration_1.default.prefix + "/package#",
    description: 'Package.',
    type: 'object',
    additionalProperties: false,
    required: ['id', 'grossWeight', 'dimensions'],
    properties: {
        id: {
            type: 'string',
            description: 'Unique identifier for package.',
        },
        type: {
            description: 'Package type.',
            $ref: packagingType_1.default.id,
        },
        grossWeight: {
            $ref: weight_1.default.id,
        },
        dimensions: {
            $ref: dimensions_1.default.id,
        },
        remainingSpace: {
            type: 'object',
            description: 'Unused space within packed package.',
            required: ['volume'],
            properties: {
                volume: {
                    $ref: volume_1.default.id,
                },
            },
        },
        items: {
            type: 'array',
            items: {
                $ref: item_1.default.id,
            },
        },
        trackingReference: {
            type: 'string',
            description: 'Per-package tracking reference. Typically supplied by carrier.',
        },
        trackingUrl: {
            description: 'URL provided for tracking individual packages.',
            type: 'string',
            format: 'uri',
        },
        documentation: {
            type: 'array',
            items: {
                $ref: document_1.default.id,
            },
        },
    },
};
