// @temando/generic-schema/schemas/product.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
exports.default = {
    $schema: 'http://json-schema.org/draft-04/schema#',
    id: configuration_1.default.prefix + "/product#",
    description: 'Product attributes.',
    type: 'object',
    additionalProperties: false,
    properties: {
        sku: {
            type: 'string',
            description: 'Product SKU.',
            maxLength: 100,
        },
        category: {
            type: 'string',
            description: 'Product category.',
        },
        description: {
            type: 'string',
            description: 'Product description.',
            maxLength: 500,
        },
        unit: {
            description: 'Product unit of measure from merchant\'s perspective. For example, kilograms, barrels, or each.',
            type: 'string',
        },
        dimensions: {
            $ref: configuration_1.default.prefix + "/dimensions#",
        },
        weight: {
            $ref: configuration_1.default.prefix + "/weight#",
        },
        monetaryValue: {
            $ref: configuration_1.default.prefix + "/money#",
        },
        isFragile: {
            description: 'Product \'is fragile\' indicator.',
            type: 'boolean',
            default: false,
        },
        origin: {
            $ref: '#/definitions/productLocation',
        },
        manufacture: {
            $ref: '#/definitions/productLocation',
        },
        classificationCodes: {
            type: 'object',
            description: 'Product export classification codes.',
            properties: {
                eccn: {
                    type: 'string',
                    description: 'Export Control Classification Number (ECCN).',
                    minLength: 0,
                    maxLength: 5,
                },
                scheduleBInfo: {
                    type: 'string',
                    description: 'Schedule B code for exporting goods out of the United States.',
                    minLength: 0,
                    maxLength: 15,
                },
                hsCode: {
                    type: 'string',
                    description: 'Harmonized Commodity Description and Coding System (HS code).',
                    minLength: 2,
                    maxLength: 15,
                },
            },
        },
    },
    definitions: {
        /**
         * This is a subset of the full location object.  A tiny subset.
         * We can turn it into a full location later.
         */
        productLocation: {
            type: 'object',
            properties: {
                address: {
                    type: 'object',
                    required: [
                        'countryCode',
                    ],
                    properties: {
                        countryCode: {
                            $ref: configuration_1.default.prefix + "/country#",
                        },
                    },
                },
            },
        },
    },
};
