// @temando/generic-schema/schemas/quote.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
var money_1 = require("./money");
var quote = {
    $schema: 'http://json-schema.org/draft-04/schema#',
    description: 'Quote.',
    type: 'object',
    additionalProperties: false,
    required: [
        'integrationId',
    ],
    properties: {
        integrationId: {
            description: 'Carrier integration ID.',
            type: 'string',
        },
        integrationServiceId: {
            description: 'Carrier service ID within the integration.',
            type: 'string',
        },
        carrierName: {
            description: 'Carrier name.',
            type: 'string',
        },
        serviceName: {
            description: 'Carrier service name.',
            type: 'string',
        },
    },
};
var ineligible = JSON.parse(JSON.stringify(quote));
ineligible.id = configuration_1.default.prefix + "/quote-ineligible#";
ineligible.description = 'Ineligible quote returned by a carrier service.';
ineligible.required = [
    'integrationId',
    'carrierName',
    'reasons',
];
ineligible.properties.reasons = {
    type: 'array',
    items: {
        description: 'Ineligibility reason.',
        type: 'string',
    },
    minItems: 1,
};
var eligible = JSON.parse(JSON.stringify(quote));
eligible.id = configuration_1.default.prefix + "/quote-eligible#";
eligible.description = 'Eligible quote returned by a carrier service.';
eligible.required = [
    'integrationId',
    'integrationServiceId',
    'carrierName',
    'serviceName',
    'shippingTaxInclusiveCharge',
    'pricedAt',
];
eligible.properties.estimatedTransitPeriod = {
    description: 'Estimated quote transit period.',
    type: 'object',
    required: [
        'minimum',
        'maximum',
    ],
    properties: {
        minimum: {
            description: 'Minimum transit time duration in ISO 8601 format.',
            $ref: configuration_1.default.prefix + "/time-duration#",
        },
        maximum: {
            description: 'Maximum transit time duration in ISO 8601 format.',
            $ref: configuration_1.default.prefix + "/time-duration#",
        },
    },
};
eligible.properties.shippingTaxInclusiveCharge = {
    $ref: money_1.default.id,
};
eligible.properties.pricedAt = {
    type: 'string',
    format: 'date-time',
};
var error = JSON.parse(JSON.stringify(quote));
error.id = configuration_1.default.prefix + "/quote-error#";
error.description = 'Error received when retrieving a quote from a carrier integration.';
error.required = [
    'integrationId',
    'carrierName',
];
error.properties.message = {
    type: 'string',
};
error.properties.errors = {
    type: 'array',
};
exports.default = {
    ineligible: ineligible,
    eligible: eligible,
    error: error,
};
