// @temando/generic-schema/schemas/shipment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_1 = require("../configuration");
exports.default = {
    $schema: 'http://json-schema.org/draft-04/schema#',
    id: configuration_1.default.prefix + "/shipment#",
    description: 'Shipment.',
    type: 'object',
    required: [
        'origin',
        'destination',
        'pickupAt',
        'packages',
    ],
    properties: {
        pickupAt: {
            type: 'string',
            description: 'Expected pick-up date-time.',
            format: 'date-time',
        },
        isDutiable: {
            description: 'Shipment \'is dutiable\' indicator.',
            type: 'boolean',
        },
        isPaperless: {
            description: 'Shipment \'is paperless\' indicator.',
            type: 'boolean',
        },
        origin: {
            $ref: configuration_1.default.prefix + "/location#",
        },
        destination: {
            $ref: configuration_1.default.prefix + "/location#",
        },
        finalRecipient: {
            $ref: configuration_1.default.prefix + "/location#",
        },
        packages: {
            type: 'array',
            items: {
                $ref: configuration_1.default.prefix + "/package#",
            },
        },
        documentation: {
            type: 'array',
            items: {
                $ref: configuration_1.default.prefix + "/document#",
            },
        },
        instructionsToDeliveryAgent: {
            type: 'string',
        },
        insuredAmount: {
            $ref: configuration_1.default.prefix + "/money#",
        },
        capabilities: {
            $ref: configuration_1.default.prefix + "/capabilities#",
        },
        exportDeclaration: {
            $ref: configuration_1.default.prefix + "/exportDeclaration#",
        },
        order: {
            type: 'object',
            description: 'Order details if shipment fulfills an order.',
            properties: {
                reference: {
                    description: 'Purchase order number associated with shipment',
                    type: 'string',
                },
                customerReference: {
                    description: 'Merchant\'s customer\'s reference.',
                    type: 'string',
                },
                invoiceId: {
                    description: 'ID of invoice associated with shipment.',
                    type: 'string',
                },
                invoiceDate: {
                    description: 'Invoice generation date-time.',
                    type: 'string',
                    format: 'date-time',
                },
            },
        },
    },
};
